html: select: refactor API

fscherf opened the pull request:
This PR refactors the `lona.html.Select` API to be more concise and convenient.

maratori commented on doc/content/end-user-documentation/html.rst:
```diff
@@ -633,40 +633,66 @@ Select
+on value, selection state and disabled state.
+
+``Option`` objects consist of a label text and a value. The value can be
+anything. If ``Option.render_value`` is set, which is set by default, the
```
Don't see `render_value` in code

maratori commented on lona/html/data_binding/select.py:
```diff
@@ -80,73 +154,109 @@ def readonly(self, new_value):
             if not self.multiple:
-                value = value.pop()
+                if not values:
+                    return self.options[0].value
```
There may be no options at all

maratori commented on lona/html/data_binding/select.py:
```diff
@@ -80,73 +154,109 @@ def readonly(self, new_value):
+            if not isinstance(new_value, list):
+                new_value = [new_value]
+
+            old_values = self.values
```
nit:
```suggestion
            values = self.values
```

fscherf replied:
Ack, Fixed!

maratori commented on lona/html/data_binding/select.py:
```diff
@@ -80,73 +154,109 @@ def readonly(self, new_value):
+            option._select = None
+            self.nodes.remove(option)
+
+    def select_all(self):
```
Maybe need to throw exception for non-multi select?

fscherf replied:
Ack, Fixed!

maratori commented on lona/html/data_binding/select.py:
```diff
@@ -80,73 +154,109 @@ def readonly(self, new_value):
-            value = []
+            options = ()
+
+            for option in self.options:
```
Let's optimize a bit: `self.options` may be calculated only once in this method.

fscherf replied:
Ack, Fixed!

maratori commented on lona/html/data_binding/select.py:
```diff
@@ -80,73 +154,109 @@ def readonly(self, new_value):
+                    raise RuntimeError(f'unknown value: {value}')
 
+            for option in self.options:
+                option._set_selected(option.value in new_value)
```
What if there are several options with the same value? I think we should select only the first one (as in browser).

fscherf replied:
Yes i agree. Like in your test

fscherf commented on lona/html/parsing.py:
```diff
@@ -143,7 +143,7 @@ def handle_starttag(self, tag, attrs, self_closing=False):
 
         # setup node
-        for key in ('id', 'class', 'style'):
+        for key in ('id', 'class', 'style', 'value'):
```
@maratori: That means:

```python
select = Select(
    Option('Foo', value='foo', selected=True),
    multiple=False,
)

option = Option('Bar', value='bar', selected=True)

select.add_option(option)

assert option.selected == False
```

Right?

I agree that this behavior would be consistent with the API i proposed, but maybe that's a little bit to much magic because  what should we do in cases like these:

```python
select = Select(
    Option('Foo', value='foo', selected=True),
    Option('Bar', value='bar', selected=True),
)

select.multiple = True

assert select.options[0].selected == True
assert select.options[1].selected == True  # fails
```

In the browser that would work. Do we want this behavior? Or do we want `Option.selected` to be dumb and the "clever" metrics like "return the first selected option if multiple are selected and it's no multi-select" are in `Select.value`, `Select.selected_values` ?

maratori replied:
> Right?

This seems to be not related to my original comment.

To answer your great question, I need to take some time to play around with this API and with browser.

fscherf replied:
> This seems to be not related to my original comment.

No? If i read your example correctly, you set two options to selected in the HTML string but assert only one `selected == True` (b and c). Am i missing something?

maratori replied:
I haven't run test, but from reading the code I think parsing string HTML doesn't work correctly with new select API. That's because `Select.append` behave differently from `Select.add_option`.

fscherf replied:
Oh! You are right. Maybe a second reason to strip the proposed `Option.selected` API

fscherf commented:
RFC: I added a patch that removes the vast majority of magic i added before. The more i think about this i think less is more here, and it is least surprising to just emulate what the browser does here.

`Option.selected` just sets or unsets `selected` in `Option.attributes` now and does not interact with the surrounding `Select`. If you decide to use a non-multiple select, but select multiple options, you can do that, but `Select.value` will return the value of the first selected option.

@maratori, @SmithChart, @grigolet: Which approach do you like better?

fscherf commented:
I created #329, on top of `master`, that contains all feedback up to this point, and using my second implementation as base. This way we make the new implementation usable before Lona 2, making the transition easier.
